File: ruby/triangle/triangle.rb
# TriangleTest class comment
class Triangle
  VERSION = 1
  
  def initialize(sides)
    @sides = sides
  end

  def equilateral?
    valid? && num_uniq == 1
  end

  def isosceles?
    valid? && num_uniq <= 2
  end

  def scalene?
    valid? && num_uniq == 3
  end

  private
  attr_reader :sides

  def valid?
    @valid ||= sides.none?(&:zero?) && (sides.sum - sides.max) >= sides.max
  end

  def num_uniq
    @num_uniq ||= sides.uniq.count
  end
end

File: python/triangle/triangle.py
def equilateral(sides):
    return len(set(sides)) == 1 and set(sides) != set([0])


def isosceles(sides):
    return (len(set(sides)) == 2 or equilateral(sides)) and not invalid_triangle(sides)


def scalene(sides):
    return len(set(sides)) == 3 and not invalid_triangle(sides)


def invalid_triangle(sides):
    a = sides[0]
    b = sides[1]
    c = sides[-1]
    return a + b <= c or a + c <= b or b + c <= a


def version():
    VERSION = 1

